network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/failure/ServerDrainingMonitor.kt

```diff
@@ -0,0 +1,114 @@
+import com.duckduckgo.networkprotection.impl.di.ProtectedVpnControllerService
+import com.squareup.anvil.annotations.ContributesMultibinding
+import com.wireguard.crypto.KeyPair
+import javax.inject.Inject
+import kotlin.time.Duration.Companion.minutes
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import logcat.logcat
+
+@ContributesMultibinding(VpnScope::class)
+class ServerDrainingMonitor @Inject constructor(
+    @ProtectedVpnControllerService private val wgVpnControllerService: WgVpnControllerService,
+    private val wgTunnelConfig: WgTunnelConfig,
+    private val networkProtectionState: NetworkProtectionState,
+    private val dispatcherProvider: DispatcherProvider,
+    private val wgTunnel: WgTunnel,
+) : VpnServiceCallbacks {
+    private val job = ConflatedJob()
+
+    override fun onVpnStarted(coroutineScope: CoroutineScope) {
+        job += coroutineScope.launch(dispatcherProvider.io()) {
+            startMonitor()
+        }
+    }
+
+    override fun onVpnReconfigured(coroutineScope: CoroutineScope) {
+        job += coroutineScope.launch(dispatcherProvider.io()) {
+            startMonitor()
+        }
+    }
+
+    override fun onVpnStopped(
+        coroutineScope: CoroutineScope,
+        vpnStopReason: VpnStopReason,
+    ) {
+        stopMonitor()
+    }
+
+    private suspend fun startMonitor() = withContext(dispatcherProvider.io()) {
+        if (networkProtectionState.isEnabled()) {
+            val serverName = wgTunnelConfig.getWgConfig()?.asServerDetails()?.serverName
+            while (isActive && networkProtectionState.isEnabled()) {
+                delay(5.minutes.inWholeMilliseconds)
+
+                if (serverName != null) {
+                    kotlin.runCatching {
+                        logcat { "Server drain monitor: getServerStatus for $serverName" }
+                        wgVpnControllerService.getServerStatus(serverName).also {
+                            if (it.shouldMigrate) {
+                                // TODO: emit pixel
+                                logcat { "Server drain monitor: attempting to migrate server" }
+                                attemptServerMigration()
+                            }
+                        }
+                    }.onFailure {
+                        logcat { "Server drain monitor: getServerStatus error $it" }
+                    }
+                }
+            }
+        } else {
+            job.cancel()
+        }
+    }
+
+    private suspend fun attemptServerMigration() {
+        kotlin.runCatching {
+            // TODO: emit pixel
+            val config = wgTunnel.createAndSetWgConfig(KeyPair())
+            networkProtectionState.restart()
+            config.getOrNull()
+        }.onFailure {
+            // TODO: emit pixel
+            logcat { "Server drain monitor: server migration failed" }
```

Comment:
Now that i have updated the way to reconnect, i don’t think this `runcatching` will effectively catch a recovery success and failure since `networkProtectionState.restart()` will most likely not throw an error here. Will find a better way.